icu4c/source/common/ulocbuilder.cpp

```diff
@@ -0,0 +1,173 @@
+// © 2023 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+
+#include <utility>
+
+#include "unicode/bytestream.h"
+#include "unicode/localebuilder.h"
+#include "unicode/locid.h"
+#include "unicode/stringpiece.h"
+#include "unicode/umachine.h"
+#include "unicode/ulocbuilder.h"
+#include "cstring.h"
+
+using icu::CheckedArrayByteSink;
+using icu::StringPiece;
+
+ULocaleBuilder* external(icu::LocaleBuilder* internal) {
+  return (ULocaleBuilder*)(internal);
+}
+
+const icu::LocaleBuilder* internal(const ULocaleBuilder* external) {
+  return (const icu::LocaleBuilder*)(external);
+}
+
+icu::LocaleBuilder* internal(ULocaleBuilder* external) {
+  return (icu::LocaleBuilder*)(external);
+}
+
+ULocaleBuilder* ulocbld_open() {
+  return external(new icu::LocaleBuilder());
+}
+
+void ulocbld_close(ULocaleBuilder* builder) {
+  if (builder == nullptr) return;
+  delete internal(builder);
+}
+
+void ulocbld_setLocale(ULocaleBuilder* builder, const char* locale, int32_t length) {
+  if (builder == nullptr) return;
+  icu::Locale l;
+  if (length < 0 || locale[length] == '\0') {
+    l = icu::Locale(locale);
+  } else {
+    if (length >= ULOC_FULLNAME_CAPACITY) {
+      l.setToBogus();
+    } else {
+      // locale is not null termined but Locale API require one.
+      // Create a null termined version in buf.
+      char buf[ULOC_FULLNAME_CAPACITY];
+      uprv_memcpy(buf, locale, length);
+      buf[length] = '\0';
+      l = icu::Locale(buf);
+    }
+  }
+  internal(builder)->setLocale(l);
+}
+
+void ulocbld_setLanguageTag(ULocaleBuilder* builder, const char* tag, int32_t length) {
+  if (builder == nullptr) return;
+  internal(builder)->setLanguageTag(
+      length < 0 ? StringPiece(tag) : StringPiece(tag, length));
```

Comment:
Didn't @markusicu say something this morning about StringPiece having logic on it that'd handle checking for the negative length for you?  Or am I wrong about this?

Reply:
The API doc said 
https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/classicu_1_1StringPiece.html#af97f22b270d21d5c97337d5d58966a59



len | the length of the string; must be non-negative
-- | --


my understanding is if you use the single parameter one it will figure out the length for you by assuming null termination. If you use the one with length  then the length will be used and no need for null termination. I cannot find one as what you remember.

Reply:
Bleah.  That's unfortunate.  But okay; I just wanted to make sure.